path_selection/path_selection.go

```diff
@@ -0,0 +1,275 @@
+
+import (
+	"errors"
+	"fmt"
+	mathrand "math/rand"
+	"time"
+
+	"github.com/katzenpost/core/crypto/ecdh"
+	"github.com/katzenpost/core/crypto/rand"
+	"github.com/katzenpost/core/epochtime"
+	"github.com/katzenpost/core/pki"
+	"github.com/katzenpost/core/sphinx"
+	"github.com/katzenpost/core/sphinx/commands"
+	"github.com/katzenpost/core/sphinx/constants"
+)
+
+// durationFromFloat returns millisecond time.Duration given a float64
+func durationFromFloat(delay float64) (time.Duration, error) {
+	d, err := time.ParseDuration(fmt.Sprintf("%fms", delay))
+	return d, err
+}
+
+// getDelays returns a list of delays from
+// the Poisson distribution with a given
+// lambda argument. As per section
+// "5.1 Choosing Delays: for single Block messages and for multi Block messages"
+// of the "Panoramix Mix Network End-to-end Protocol Specification"
+// the delay for the egress provider, the last hop is always zero,
+// see https://github.com/Katzenpost/docs/blob/master/specs/end_to_end.txt
+func getDelays(lambda float64, count int) []float64 {
+	cryptRand := rand.NewMath()
+	delays := make([]float64, count)
+	for i := 0; i < count-2; i++ {
+		delays[i] = rand.Exp(cryptRand, lambda)
+	}
+	delays[count-1] = float64(0)
+	return delays
+}
+
+// sum adds a slice of float64.
+// this is used to get the sum of delays
+// which are represented as float64s
+func sum(input []float64) float64 {
+	total := float64(0)
+	for _, n := range input {
+		total += n
+	}
+	return total
+}
+
+// RouteFactory builds routes and by doing so handles all the
+// details of path selection and mix PKI interaction. This factory
+// is used to create routes for a mix network which uses the
+// Poisson mix strategy, where the client selects a delay for each
+// hop in the route where the mean of this distribution is tuneable
+// using the lambda parameter.
+type RouteFactory struct {
+	pki    pki.Client
+	nrHops int
+	lambda float64
+}
+
+// NewRouteFactory creates a new RouteFactory for creating routes
+func New(pki pki.Client, nrHops int, lambda float64) *RouteFactory {
+	r := RouteFactory{
+		pki:    pki,
+		nrHops: nrHops,
+		lambda: lambda,
+	}
+	return &r
+}
+
+// getRouteDescriptors returns a slice of mix descriptors,
+// one for each hop in the route where each mix descriptor
+// was selected from the set of descriptors for that layer
+func (r *RouteFactory) getRouteDescriptors(senderProviderName, recipientProviderName string) ([]*pki.MixDescriptor, error) {
+	var err error
+	descriptors := make([]*pki.MixDescriptor, r.nrHops)
+	descriptors[0], err = r.pki.GetProviderDescriptor(senderProviderName)
+	if err != nil {
+		return nil, err
+	}
+	descriptors[r.nrHops-1], err = r.pki.GetProviderDescriptor(recipientProviderName)
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < r.nrHops; i++ {
+		layerMixes := r.pki.GetMixesInLayer(uint8(i))
+		fmt.Printf("layer %d\n", i)
+		if len(layerMixes) == 0 {
+			panic("WTF")
+		}
+		c := mathrand.Intn(len(layerMixes))
```

Comment:
are you sure mathrand is okay? i'd default to a crypto/rand.Int if i wasn't sure.

Reply:
ah indeed that's a better choice